### PR TITLE
Add Reactive Support for UserDetailsChecker

### PR DESCRIPTION
**Pull request by edeandrea**
Integrate UserDetailsChecker into ReactiveAuthenticationManager and
OAuth2 resource server authentication converters.

Fixes gh-6219

### CODE REVIEW

File: core/src/main/java/org/springframework/security/authentication/UserDetailsRepositoryReactiveAuthenticationManager.java [hunk 1]
```diff
@@ -43,6 +44,8 @@
 
 	private Scheduler scheduler = Schedulers.parallel();
 
+	private UserDetailsChecker userDetailsChecker;
```
Comment (rwinch) [inline core/src/main/java/org/springframework/security/authentication/UserDetailsRepositoryReactiveAuthenticationManager.java]: We should default this to just do nothing, so we don't need to worry about null values.

File: core/src/main/java/org/springframework/security/authentication/UserDetailsRepositoryReactiveAuthenticationManager.java [hunk 2]
```diff
@@ -102,4 +110,16 @@ public void setUserDetailsPasswordService(
...
 	}
+
+	/**
+	 * Sets the strategy which will be used to validate the loaded <tt>UserDetails</tt>
+	 * object for the user.
+	 *
+	 * @param userDetailsChecker The {@link UserDetailsChecker}
+	 * @since 5.2
+	 */
+	public void setUserDetailsChecker(UserDetailsChecker userDetailsChecker) {
```
Comment (rwinch) [inline core/src/main/java/org/springframework/security/authentication/UserDetailsRepositoryReactiveAuthenticationManager.java]: We should expose this as `setPostAuthenticationChecks` to align with `AbstractUserDetailsAuthenticationProvider`. We should also update the member variable to `postAuthenticationChecks`

File: oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/JwtUserDetailsCheckingAuthenticationConverter.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
+/*
```
Comment (rwinch) [inline oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/JwtUserDetailsCheckingAuthenticationConverter.java]: Please remove this from the PR as it is something that should be considered separately.

File: oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/ReactiveJwtUserDetailsCheckingAuthenticationConverter.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
+/*
```
Comment (rwinch) [inline oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/ReactiveJwtUserDetailsCheckingAuthenticationConverter.java]: Please remove this from the PR as it is something that should be considered separately.

File: oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/authentication/JwtUserDetailsCheckingAuthenticationConverterTests.java [hunk 5]
```diff
@@ -0,0 +1,238 @@
+/*
```
Comment (rwinch) [inline oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/authentication/JwtUserDetailsCheckingAuthenticationConverterTests.java]: Please remove this from the PR as it is something that should be considered separately.

File: oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/authentication/ReactiveJwtUserDetailsCheckingAuthenticationConverterTests.java [hunk 6]
```diff
@@ -0,0 +1,268 @@
+/*
```
Comment (rwinch) [inline oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/authentication/ReactiveJwtUserDetailsCheckingAuthenticationConverterTests.java]: Please remove this from the PR as it is something that should be considered separately.

File: core/src/main/java/org/springframework/security/authentication/UserDetailsRepositoryReactiveAuthenticationManager.java (hunk 1 above)
Comment (edeandrea) [reply core/src/main/java/org/springframework/security/authentication/UserDetailsRepositoryReactiveAuthenticationManager.java]: Do you mean build some kind of `NoOpUserDetailsChecker` and default to that? Or do something like this inside `UserDetailsRepositoryReactiveAuthenticationManager`

```java
private UserDetailsChecker userDetailsChecker = userDetails -> {};
```

Comment (rwinch) [reply core/src/main/java/org/springframework/security/authentication/UserDetailsRepositoryReactiveAuthenticationManager.java]: Just use a lambda

**Comment by edeandrea** [comment]
I made the requested changes & rebased the commits into a single commit. I'll open a separate ticket for the `UserDetailsCheckingAuthenticationConverter`s.

**Comment by edeandrea** [comment]
I opened #6237 for the other part of what I removed from this PR.

File: core/src/test/java/org/springframework/security/authentication/UserDetailsRepositoryReactiveAuthenticationManagerTests.java [hunk 7]
```diff
@@ -56,6 +55,9 @@
 	@Mock
 	private Scheduler scheduler;
 
+	@Mock
+	private UserDetailsChecker userDetailsChecker;
```
Comment (rwinch) [inline core/src/test/java/org/springframework/security/authentication/UserDetailsRepositoryReactiveAuthenticationManagerTests.java]: Can you please rename this to `postAuthenticationChecks` to be consistent?

File: core/src/test/java/org/springframework/security/authentication/UserDetailsRepositoryReactiveAuthenticationManagerTests.java [hunk 8]
```diff
@@ -140,4 +142,19 @@ public void authenticateWhenPasswordServiceAndUpgradeFalseThenNotUpdated() {
...
+		this.manager.setPasswordEncoder(this.encoder);
+		this.manager.setPostAuthenticationChecks(this.userDetailsChecker);
+
+		assertThatExceptionOfType(LockedException.class)
+				.isThrownBy(() -> this.manager.authenticate(new UsernamePasswordAuthenticationToken(this.user, this.user.getPassword())).block())
+				.withMessage("account is locked");
+
+		verify(this.userDetailsChecker).check(eq(this.user));
+	}
 }
```
Comment (rwinch) [inline core/src/test/java/org/springframework/security/authentication/UserDetailsRepositoryReactiveAuthenticationManagerTests.java]: Can you add another test that verifies that `postAuthenticationChecks` is not invoked if the `UserDetails` is not found?

File: core/src/test/java/org/springframework/security/authentication/UserDetailsRepositoryReactiveAuthenticationManagerTests.java (hunk 7 above)
Comment (edeandrea) [reply core/src/test/java/org/springframework/security/authentication/UserDetailsRepositoryReactiveAuthenticationManagerTests.java]: Done!

File: core/src/test/java/org/springframework/security/authentication/UserDetailsRepositoryReactiveAuthenticationManagerTests.java (hunk 8 above)
Comment (edeandrea) [reply core/src/test/java/org/springframework/security/authentication/UserDetailsRepositoryReactiveAuthenticationManagerTests.java]: Done!

**Comment by edeandrea** [comment]
Added a new commit with requested changes.

**Comment by edeandrea** [comment]
Added a new commit with requested changes.

**Comment by edeandrea** [comment]
Done!

**Pull request merged**